Yank/analyze.py

```diff
@@ -32,27 +35,1145 @@
+
+    @reference_states.setter
+    def reference_states(self, value):
+        """Provide a way to re-assign the i,j states in a protected way"""
+        i, j = value[0], value[1]
+        if type(i) is not int or type(j) is not int:
+            raise ValueError("reference_states must be a length 2 iterable of ints")
+        self._reference_states = (i, j)
+
+    @property
+    def kT(self):
+        """Fetch the kT of the phase"""
+        if self._kT is None:
+            thermodynamic_states, _ = self._reporter.read_thermodynamic_states()
+            temperature = thermodynamic_states[0].temperature
+            self._kT = kB * temperature
+        return self._kT
+
+    @property
+    def reporter(self):
+        """Return the reporter tied to this object..."""
+        return self._reporter
+
+    @reporter.setter
+    def reporter(self, value):
+        """... and then make sure users cannot overwrite it."""
+        raise ValueError("You cannot re-assign the reporter for this analyzer!")
+
+    # Abstract methods
+    @abc.abstractmethod
+    def analyze_phase(self, *args, **kwargs):
+        """
+        Function which broadly handles "auto-analysis" for those that do not wish to call all the methods on their own.
+        This should be have like the old "analyze" function from versions of YANK pre-1.0.
+
+        Returns a dictionary of analysis objects
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _create_mbar_from_scratch(self):
+        """
+        This method should automatically do everything needed to make the MBAR object from file. It should make all
+        the assumptions needed to make the MBAR object.  Typically alot of these functions will be needed for the
+        analyze_phase function,
+
+        Returns nothing, but the self.mbar object should be set after this
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def extract_energies(self):
+        """
+        Extract the deconvoluted energies from a phase. Energies from this are NOT decorrelated.
+
+        Returns
+        -------
+        sampled_energy_matrix : Deconvoluted energy of sampled states evaluated at other sampled states.
+            Has shape (K,K,N) = (number of sampled states, number of sampled states, number of iterations)
+            Indexed by [k,l,n]
+            where an energy drawn from sampled state [k] is evaluated in sampled state [l] at iteration [n]
+        unsampled_energy_matrix
+            Has shape (K, L, N) = (number of sampled states, number of UN-sampled states, number of iterations)
+            Indexed by [k,l,n]
+            where an energy drawn from sampled state [k] is evaluated in un-sampled state [l] at iteration [n]
+        """
+        raise NotImplementedError()
+
+    # This SHOULD be an abstract static method since its related to the analyzer, but could handle any input data
+    # Until we drop Python 2.7, we have to keep this method
```

Comment:
Python 2 does not support
```python
@staticmethod
@abc.abstractmethd
def function():
...
```
It raises an error... for... reasons.

We did not talk about it.